### PR TITLE
`Assessment`: Support long feedback messages

### LINKED ISSUE #6296: Long Feedback Texts
### Is your feature request related to a problem?

Similar to #4796, we have some exercises that test programs via CLI interactions. Therefore, more text output is generated which sometimes is longer than 5.000 characters.

## Data Model

A new database table `long_feedback` is created (pseudo-SQL):
```sql
create table long_feedback (
   id serial primary key,
   feedback_text longtext, -- liquibase type gets mapped to longtext in MySQL , text in PostgreSQL
   feedback serial foreign key references feedback.id
)
```

The existing `feedback` gets extended with a `bool` flag `long_feedback_exists` that indicates whether this feedback was truncated and a separate long feedback is stored.
Therefore, each long feedback is linked to exactly one regular feedback and cannot exist without one. The `feedback` only contains information if some optional long feedback exists.

### Server 

New feedback creation: Check if feedback is longer than the current limit:
- if not, no special action is needed
- otherwise:
  - truncate the regular feedback text to either the first line, or to the first 200 (?) characters similar to how it is displayed in the client currently and set the long feedback flag
  - save the full feedback text separately into the long feedback table and connect it to the regular feedback it belongs to

Endpoint to fetch feedback:
- When delivering results/feedbacks to the client, the long feedbacks are not transmitted. The client can determine later due to the flag if additional long feedback should be fetched separately.
- A new REST endpoint `api/participations/{participationId}/results/{resultId}/feedback/{feedbackId}/long-feedback` is introduced.
  - This delivers the `feedback_text` of the `LongFeedback` connected to the regular feedback with `feedbackId` to the client as plain text.

### Client

When feedback is displayed, only the flag if long feedback exists is known.
Based on this, the information that the feedback has been truncated can be displayed. Additionally, a link should be present that opens a new tab where the long feedback text is shown as plain text information.
Showing the feedback inside the Artemis UI would have little benefit, since there is no way to syntax-highlight relevant parts, so it would have to be displayed there as a text blob anyway.

## Security

Even the feedback will still be truncated at some reasonable limit that should realistically never be exceeded, e.g. a few MiBs (that would still allow feedback with a few million text characters). This prevents denial-of-service attacks where arbitrarily large feedback has to be processed and stored. E.g. students might be able to craft long logs/assertion message content in their programs that would then be stored as feedback in Artemis.

### Describe alternatives you've considered

_No response_

### Additional context

In case the full feature is not ready for the maintenance window at TUM (March 6th to 10th), a minimal database migration could be applied to the large `feedback` table to make use of the downtime there and avoid separate Artemis service interruptions:
The `feedback` table gets extended with the `bool` flag `long_feedback_exists` which then would remain unused until the full implementation is ready.

### PR DESCRIPTION
**Pull request by b-fein**
# Contains database migrations

#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/guidelines/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
#### Changes affecting Programming Exercises
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite).
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).

### Motivation and Context
Closes #6296.

### Description
- Adds a new table for the long feedback stored separately from the regular feedback to allow for lazy-loading the long feedback only if actually needed.
- The client only loads the long feedback in case the actual feedback text is un-collapsed. The latest requests are also cached, so that opening and closing the feedback item does not repeatedly fetch the data from the server again.
- For very long feedback a download link to get the full feedback is shown together with the truncated version.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Programming Exercise

Stacked PR on top of #6384. If you can still access the online editor, the changes from #6384 have also been successfully tested.

1. As instructor, create a test case that produces a feedback longer than 5000 characters. E.g.
   ```java
   @Test
   void longFeedback() {
       assertEquals("a".repeat(1000), "1234567890".repeat(500));
   }
   ```
2. Check the result for the template or solution repository. The feedback should contain the full message, i.e. it ends with `…1234567890>` instead of being cut off somewhere in the middle of the sequence.
3. Increase the `500` of the repeat to something like `2000`. Now the feedback is too long to be displayed fully in the UI, instead a download link should be shown.
4. In the network tab of the browser dev console: When you un-collapse the feedback message for the first time, a network request to `long-feedback` should be made. After collapsing, or even closing the feedback detail dialogue and opening it again, the request is not made again.

*Note*: The feedback message for Gradle exercises is cut in half on Jenkins setups due to an unrelated bug currently. → #6434

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Branch | Line |
|------------|-------:|-----:|
| `Feedback.java` | 83% | 97% |
| `LongFeedbackTextResource.java` | 100% | 100% |
| `feedback-text.component.ts` | 100% | 100% |

### Screenshots

Screenshot 2023-03-26 at 11-52-31 Code Editor

Download link for very long feedback:
Screenshot 2023-03-26 at 11-52-44 Code Editor

### CODE REVIEW

File: src/main/webapp/app/exercises/shared/feedback/feedback-service.ts [hunk 1]
```diff
@@ -28,6 +28,11 @@ export class FeedbackService {
      * @param result
      */
     public getFeedbacksForResult(participationId: number, result: Result): Observable<Feedback[]> {
-        return this.resultService.getFeedbackDetailsForResult(participationId, result).pipe(map(({ body: feedbackList }) => feedbackList!));
+        return this.resultService.getFeedbackDetailsForResult(participationId, result).pipe(
```
Comment (simon-hng) [inline src/main/webapp/app/exercises/shared/feedback/feedback-service.ts:31]: I think this is not necessary since it is already done in the resultService

```ts
    getFeedbackDetailsForResult(participationId: number, result: Result): Observable<HttpResponse<Feedback[]>> {
        return this.http.get<Feedback[]>(`${this.participationResourceUrl}/${participationId}/results/${result.id!}/details`, { observe: 'response' }).pipe(
            map((res) => {
                const feedbacks = res.body ?? [];
                feedbacks.forEach((feedback) => (feedback.result = result));
                return res;
            }),
        );
    }```

File: src/main/webapp/app/exercises/shared/feedback/feedback.component.ts [hunk 2]
```diff
@@ -180,6 +180,8 @@ export class FeedbackComponent implements OnInit {
                 switchMap((feedbacks: Feedback[] | undefined | null) => {
                     if (feedbacks && feedbacks.length) {
                         this.result.feedbacks = feedbacks!;
+                        feedbacks.forEach((feedback) => (feedback.result = this.result));
```
Comment (simon-hng) [inline src/main/webapp/app/exercises/shared/feedback/feedback.component.ts:183]: Why is this necessary? The feedback results were set in `resultService.getFeedbackDetailsForResult`, `feedbackService.getFeedbacksForResult` (change above) and now here

File: src/main/webapp/app/exercises/shared/feedback/item/feedback-item.ts [hunk 3]
```diff
@@ -11,4 +12,17 @@ export class FeedbackItem implements FeedbackNode {
     text?: string; // this is typically feedback.detailText
     positive?: boolean;
     color?: FeedbackColor;
+    feedbackReference: FeedbackReference;
+}
+
+export class FeedbackReference {
+    feedbackId: number;
```
Comment (simon-hng) [inline src/main/webapp/app/exercises/shared/feedback/item/feedback-item.ts:19]: Cant we just keep the entire `feedback`, `result` as reference instead of only the ID?

**Review by simon-hng — COMMENTED** [review]
Left some questions

File: src/main/webapp/app/exercises/shared/feedback/item/feedback-item.ts (hunk 3 above)
Comment (b-fein) [reply src/main/webapp/app/exercises/shared/feedback/item/feedback-item.ts]: Yes, we can. Adapted it.
The original intention was to keep the stored information in `feedbackItem` to a minimum with only non-`undefined` values. I guess the IDs should always be present anyway and an `undefined` `hasLongFeedbackText` can safely be assumed to be `false`.

File: src/main/webapp/app/exercises/shared/feedback/feedback.component.ts (hunk 2 above)
Comment (b-fein) [reply src/main/webapp/app/exercises/shared/feedback/feedback.component.ts]: You are right, it is only needed if the feedbacks are retrieved via the `result` input to the component. In that case they might not be connected properly. I moved this line up into the relevant case.

File: src/main/webapp/app/exercises/shared/feedback/feedback-service.ts (hunk 1 above)
Comment (b-fein) [reply src/main/webapp/app/exercises/shared/feedback/feedback-service.ts]: :+1: removed

**Review by chrisknedl — DISMISSED** [review]
Tested locally on my Jenkins-Gitlab setup. Everything works as intended and the API call is also made.

**Comment by pal03377** [comment]
After adding the test case with the short message and going back to the solution results, I see this (TS2):

screenshot-2023-04-11_001222

**Review by TimOrtel — APPROVED** [review]
Tested on ts2, works for java maven but not for gradle. Allows me to download long feedback text as a txt file.

**Review by danielbode — APPROVED** [review]
Tested on ts2 and it works for Maven Java exercises.

**Review by frederik-kt — APPROVED** [review]
Tested on TS2. Works as expected for Java Maven.

**Review by tobias-lippert — APPROVED** [review]
Works for maven. The download link is displayed and the file can be downloaded and contains the entire test feedback.

File: src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseIntegrationTestService.java [hunk 4]
```diff
@@ -75,6 +75,8 @@
 @Service
 class ProgrammingExerciseIntegrationTestService {
 
+    private static final String NON_EXISTING_ID = Integer.toString(1_000_000);
+
```
Comment (chrisknedl) [inline src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseIntegrationTestService.java:79]:
```suggestion

    private static final String NON_EXISTING_ID = Integer.toString(Integer.MAX_VALUE);

```
Why not use Integer.MAX_VALUE here? If the IDs are auto-generated, that should be even more future-proof.

**Review by chrisknedl — COMMENTED** [review]
Thanks for implementing this feature. I looked at all the changes and left some suggestions for (mostly cosmetic) things that I found.

File: src/main/java/de/tum/in/www1/artemis/config/Constants.java [hunk 5]
```diff
@@ -125,6 +125,8 @@ public final class Constants {
 
     public static final int FEEDBACK_DETAIL_TEXT_MAX_CHARACTERS = 5000;
 
+    public static final int LONG_FEEDBACK_MAX_LENGTH = 10_000_000;
```
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/config/Constants.java:128]:
```suggestion
    /**
     * Arbitrary limit that is unlikely to be reached by real feedback in practice.
     * Avoids filling the DB with huge text blobs, e.g. in case an infinite loop in a test case outputs a lot of text.
     */
    public static final int LONG_FEEDBACK_MAX_LENGTH = 10_000_000;
```

File: src/main/java/de/tum/in/www1/artemis/domain/LongFeedbackText.java [hunk 6]
```diff
@@ -30,8 +65,39 @@ public void setFeedback(Feedback feedback) {
...
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        }
+        else if (obj instanceof LongFeedbackText longFeedbackText) {
+            final boolean bothIdsNull = getId() == null && longFeedbackText.getId() == null;
+            return !bothIdsNull && Objects.equals(getId(), longFeedbackText.getId());
+        }
+        else {
```
Comment (simon-hng) [inline src/main/java/de/tum/in/www1/artemis/domain/LongFeedbackText.java:89]: redundant else return

**Review by simon-hng — APPROVED** [review]
Code looks good. Thanks for implementing my suggestions

**Review by pal03377 — DISMISSED** [review]
> > After adding the test case with the short message and going back to the solution results, I see this (TS2)
> 
> @pal03377 Thanks for testing! Was that per chance an exercise that was used to test #6284/#6414 before? (test user 12 does not have access to the course it is in, or the exercise you linked no longer exists) In that case it fails since the build plan can no longer be fetched from Artemis after deploying this PR.

@b-fein 
It was! In this case I approve. It worked on other exercises 👍

File: src/main/java/de/tum/in/www1/artemis/domain/Feedback.java [hunk 7]
```diff
@@ -127,27 +137,65 @@ public Feedback detailText(@Nullable String detailText) {
     }
 
     /**
-     * sets the detail text of the feedback. In case the detail text is longer than 5000 characters, the additional characters are cut off to avoid database issues
+     * Sets the detail text of the feedback.
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/domain/Feedback.java:140]: I would mention the rough logic of the method here, in particular if feedback is larger than FEEDBACK_DETAIL_TEXT_MAX_CHARACTERS, only FEEDBACK_DETAIL_TEXT_MAX_CHARACTERS will be stored in the detailText and reference to a new object LongFeedbackText with the full / remaining characters will be created

File: src/main/java/de/tum/in/www1/artemis/config/Constants.java [hunk 8]
```diff
@@ -18,9 +18,9 @@ public final class Constants {
 
     public static int COMPLAINT_LOCK_DURATION_IN_MINUTES = 24 * 60; // 24h; Same as in artemisApp.locks.acquired
 
-    public static int SECONDS_BEFORE_RELEASE_DATE_FOR_COMBINING_TEMPLATE_COMMITS = 15;
+    public static final int SECONDS_BEFORE_RELEASE_DATE_FOR_COMBINING_TEMPLATE_COMMITS = 15;
 
-    public static int SECONDS_AFTER_RELEASE_DATE_FOR_UNLOCKING_STUDENT_EXAM_REPOS = 5;
+    public static final int SECONDS_AFTER_RELEASE_DATE_FOR_UNLOCKING_STUDENT_EXAM_REPOS = 5;
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/config/Constants.java:23]: Why did you change these two but not any other variable to final?

**Review by JohannesStoehr — DISMISSED** [review]
Code looks good to me. Left one optional comment

File: src/main/java/de/tum/in/www1/artemis/config/Constants.java (hunk 8 above)
Comment (b-fein) [reply src/main/java/de/tum/in/www1/artemis/config/Constants.java:23]: The one above is actually modified in the tests to get a shorter timeout. The others were already `final` as they should be.

**Review by krusche — APPROVED** [review]
